mpisppy/convergers/primal_dual_converger.py

```diff
@@ -0,0 +1,118 @@
+            for ndn_i, xbar in s._mpisppy_model.xbars.items():
+                xbars[ndn_i] = xbar.value
+        return xbars
+    
+    def _compute_primal_convergence(self):
+        """
+        Compute the primal convergence metric
+        Returns:
+            global_sum_diff (float): primal convergence metric
+        """
+        local_sum_diff = np.zeros(1)
+        global_sum_diff = np.zeros(1)
+        for _, s in self._ph.local_scenarios.items():
+            # we iterate over decision nodes instead of
+            # s._mpisppy_data.nonant_indices to use numpy
+            for node in s._mpisppy_node_list:
+                ndn = node.name
+                nlen = s._mpisppy_data.nlens[ndn]
+                x_bars = np.fromiter((s._mpisppy_model.xbars[ndn,i]._value
+                                      for i in range(nlen)), dtype='d')
+                
+                nonants_array = np.fromiter(
+                    (v._value for v in node.nonant_vardata_list),
+                    dtype='d', count=nlen)
+                _l1 = np.abs(x_bars - nonants_array)
+                
+                # invariant to prob_coeff being a scalar or array
+                prob = s._mpisppy_data.prob_coeff[ndn] * np.ones(nlen)
+                local_sum_diff[0] += np.dot(prob, _l1)
+
+        self._ph.comms["ROOT"].Allreduce(local_sum_diff, global_sum_diff, op=MPI.SUM)
+        return global_sum_diff[0]
+
+    def _compute_dual_residual(self):
+        """ Compute the dual residual
+
+        Returns:
+           global_diff (float): difference between to consecutive x bars
+
+        """
+        local_sum_diff = np.zeros(1)
+        global_sum_diff = np.zeros(1)
+        for s in self._ph.local_scenarios.values():
+            for node in s._mpisppy_node_list:
+                ndn = node.name
+                nlen = s._mpisppy_data.nlens[ndn]
+                rhos = np.fromiter((s._mpisppy_model.rho[ndn,i]._value
+                                    for i in range(nlen)), dtype='d')
+                xbars = np.fromiter((s._mpisppy_model.xbars[ndn,i]._value
+                                        for i in range(nlen)), dtype='d')
+                prev_xbars = np.fromiter((self.prev_xbars[ndn,i]
+                                            for i in range(nlen)), dtype='d')
+
+                local_sum_diff[0] += np.sum(rhos * np.abs(xbars - prev_xbars))
+
+        self._ph.comms["ROOT"].Allreduce(local_sum_diff, global_sum_diff, op=MPI.SUM)
+        return global_sum_diff[0]
+
+    def is_converged(self):
+        """ check for convergence
+        Args:
+            self (object): create by prep
+
+        Returns:
+           converged?: True if converged, False otherwise
+        """
+
+        primal_gap = self._compute_primal_convergence()
+        dual_gap = self._compute_dual_residual()
+        self.prev_xbars = self._get_xbars()
+        ret_val = primal_gap + dual_gap <= self.convergence_threshold
```

Comment:
Why this? And not:

```suggestion
        ret_val = max(primal_gap, dual_gap) <= self.convergence_threshold
```
?

Reply:
I don't have a good rationale, except that is what the original PySP implementation had. In practice, one of the terms is 10-100x the other one, so it is effectively a max and we can do that

Reply:
Ben makes the case that max is easier to reason about when setting the convergence threshold, so please make the change.